fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java

```diff
@@ -1493,60 +1506,78 @@ private List<TScanRangeParams> findOrInsert(Map<Integer, List<TScanRangeParams>>
             }
-            addressToScanRanges.get(address).add(scanRanges);
+            addressToScanRanges.get(address).add(bucketSeqAndScanRanges);
         }
 
         for (Map.Entry<TNetworkAddress, List<Map.Entry<Integer, Map<Integer,
                 List<TScanRangeParams>>>>> addressScanRange : addressToScanRanges.entrySet()) {
             List<Map.Entry<Integer, Map<Integer, List<TScanRangeParams>>>> scanRange = addressScanRange.getValue();
+
             int expectedInstanceNum = 1;
             if (parallelExecInstanceNum > 1) {
-                //the scan instance num should not larger than the tablets num
+                // The scan instance num should not larger than the tablets num
                 expectedInstanceNum = Math.min(scanRange.size(), parallelExecInstanceNum);
             }
 
             // 2. split how many scanRange one instance should scan
-            List<List<Map.Entry<Integer, Map<Integer, List<TScanRangeParams>>>>> perInstanceScanRanges =
-                    ListUtil.splitBySize(scanRange,
-                            expectedInstanceNum);
+            List<List<Map.Entry<Integer, Map<Integer, List<TScanRangeParams>>>>> scanRangesPerInstance =
+                    ListUtil.splitBySize(scanRange, expectedInstanceNum);
 
             // 3.construct instanceExecParam add the scanRange should be scan by instance
-            for (List<Map.Entry<Integer, Map<Integer, List<TScanRangeParams>>>> perInstanceScanRange : perInstanceScanRanges) {
+            for (List<Map.Entry<Integer, Map<Integer, List<TScanRangeParams>>>> scanRangePerInstance : scanRangesPerInstance) {
                 FInstanceExecParam instanceParam = new FInstanceExecParam(null, addressScanRange.getKey(), 0, params);
                 // record each instance replicate scan id in set, to avoid add replicate scan range repeatedly when they are in different buckets
                 Set<Integer> instanceReplicateScanSet = new HashSet<>();
-                for (Map.Entry<Integer, Map<Integer, List<TScanRangeParams>>> nodeScanRangeMap : perInstanceScanRange) {
-                    instanceParam.addBucketSeq(nodeScanRangeMap.getKey());
-                    for (Map.Entry<Integer, List<TScanRangeParams>> nodeScanRange : nodeScanRangeMap.getValue()
-                            .entrySet()) {
-                        int scanId = nodeScanRange.getKey();
-                        if (!instanceParam.perNodeScanRanges.containsKey(scanId)) {
-                            instanceParam.perNodeScanRanges.put(scanId, new ArrayList<>());
-                        }
-                        if (replicateScanIds.contains(scanId)) {
-                            if (!instanceReplicateScanSet.contains(scanId)) {
-                                instanceParam.perNodeScanRanges.get(scanId).addAll(nodeScanRange.getValue());
-                                instanceReplicateScanSet.add(scanId);
+
+                int expectedDop = 1;
```

Comment:
Could we abstract some pipeline related codes outer from Coordinator class?

Reply:
Here we assign bucket sequences of tablets to each driver sequence, which should consider pipeline DOP and driver sequence, so maybe we couldn't move it outer from `Coordinator`?